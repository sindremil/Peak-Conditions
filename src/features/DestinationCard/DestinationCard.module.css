.destinationCard {
  width: 25vw;
  height: 35vw;
  border-radius: 3vw;
  font-family: roboto-regular;
  overflow: hidden;
  display: flex;
  justify-content: left;
  flex-wrap: wrap;
}

p {
  margin: 0;
}

.destinationCardImg {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.destinationNameContainer {
  width: 100%;
  display: flex;
  justify-content: left;
}

.destinationInfoContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: baseline;
  width: 50%
}

.destinationSymbol {
  width: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.destinationName {
  font-size: 2.5vw;
  margin-left: 5%;
  margin-top: 2%;
}

.destinationTemp, .destinationWind {
  font-size: 2.2vw;
  margin-left: 10%;
}

.weatherIcon {
  width: 70%;
}

.details {
  flex-basis: 100%;
  font-size: 1.67vw;
  color: white;
  text-decoration: none;
  text-align: center;
}

.details:hover {
  background-color: #2e485c;
  cursor: pointer;
  color: white;
}

.destinationCardDivider {
  margin-top: 0;
  color: white;
}

.imgContainer {
  position: relative;
  width: 100%;
  height: 50%;
}

.favourite {
  width: 3.33vw;
  height: 3.33vw;
  cursor: pointer;
  position: absolute;
  top: 5px;
  right: 5px;
}


/* Change sizes when using mobile, so that cards appear one after one */

@media (max-aspect-ratio: 3/4) {

  .destinationCard {
    width: 100%;
    height: 126vw;
    border-radius: 10vw;
  }

  .destinationName {
    font-size: 9vw;
  }

  .destinationTemp, .destinationWind {
    font-size: 9.2vw;
  }

  .details {
    font-size: 6vw;
  }

  .favourite   {
    width: 12vw;
    height: 12vw;
  }
}