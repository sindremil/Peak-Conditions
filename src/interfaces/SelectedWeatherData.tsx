export default interface SelectedWeatherData {
  destination: string;
  temperature: number;
  windSpeed: number;
  symbolCode: string;
}
