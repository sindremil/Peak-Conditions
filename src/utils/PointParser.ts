import Point from '../interfaces/Point.tsx';

//Parses a Point object into an API query
export default function getPointQuery(point: Point): string {
  const { lat, lon, alt } = point;
  return (
    'https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=' +
    lat +
    '&lon=' +
    lon +
    '&altitude=' +
    alt
  );
}
