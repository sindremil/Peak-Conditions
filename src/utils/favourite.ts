export function getFavouritesArray(): string[] {
  const favoritesString: string | null = localStorage.getItem('favourites');
  let favoritesArray: string[] = [];
  if (favoritesString) {
    favoritesArray = JSON.parse(favoritesString);
  }
  return favoritesArray;
}

export function addFavourite(destinationName: string) {
  const favoritesArray = getFavouritesArray();
  favoritesArray.push(destinationName);
  localStorage.setItem('favourites', JSON.stringify(favoritesArray));
}

export function removeFavourite(destinationName: string) {
  const favoritesArray = getFavouritesArray();
  const destinationIndex = favoritesArray.indexOf(destinationName);

  //If destinatnion not found, then do nothing
  if (destinationIndex !== -1) {
    favoritesArray.splice(destinationIndex, 1);
  }
  localStorage.setItem('favourites', JSON.stringify(favoritesArray));
}

export function isFavourite(destinationName: string): boolean {
  const favouritesArray: string[] = getFavouritesArray();
  if (favouritesArray.indexOf(destinationName) == -1) {
    return false;
  }
  return true;
}
