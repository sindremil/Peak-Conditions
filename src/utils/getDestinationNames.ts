import destinationsConfig from "../configs/destinations.json";

export default function getDestinationNames() {
  const destinationNames = destinationsConfig.destinations.map(
    (destination) => destination.name,
  );
  return destinationNames;
}
