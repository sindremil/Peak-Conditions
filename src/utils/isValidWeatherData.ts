import WeatherData from "../interfaces/WeatherData";

export default function isValidWeatherData(weatherData: WeatherData): boolean {
  return (
    weatherData &&
    weatherData.properties &&
    weatherData.properties.timeseries &&
    weatherData.properties.timeseries.length > 0
  );
}
