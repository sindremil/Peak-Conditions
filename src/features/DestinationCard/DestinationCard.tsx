import notFavourite from '../../assets/star1.svg';
import style from './DestinationCard.module.css';
import favourite from '../../assets/star2.svg';
import { useState } from 'react';
import SelectedWeatherData from '../../interfaces/SelectedWeatherData';
import { Link } from 'react-router-dom';
import { addFavourite, removeFavourite } from '../../utils/favourite';

interface DestinationCardProps extends SelectedWeatherData {
  isLocalStorageFavourite: boolean;
}

export default function DestinationCard({
  destination,
  temperature,
  windSpeed,
  symbolCode,
  isLocalStorageFavourite,
}: DestinationCardProps) {
  const [isFavourite, setIsFavorite] = useState(isLocalStorageFavourite);

  const destinationImgPath: string =
    'images/destinations/' + destination.toLowerCase() + '.jpg';
  const symbolImgPath: string = 'images/weather/' + symbolCode + '.svg';

  function handleFavorite() {
    setIsFavorite(!isFavourite);
    if (!isFavourite) {
      addFavourite(destination);
    } else {
      removeFavourite(destination);
    }
  }

  return (
    <div className={`${'card'} ${style.destinationCard}`}>
      <div className={style.imgContainer}>
        <img
          className={style.destinationCardImg}
          src={destinationImgPath}
          alt={destination}
        />
        <img
          className={style.favourite}
          onClick={handleFavorite}
          src={isFavourite ? favourite : notFavourite}
          alt="Favourite star"
        />
      </div>
      <div className={style.destinationNameContainer}>
        <p className={style.destinationName}>{destination}</p>
      </div>
      <div className={style.destinationInfoContainer}>
        <p className={style.destinationTemp}>{temperature}Â°</p>
        <p className={style.destinationWind}>{windSpeed}m/s</p>
      </div>
      <div className={style.destinationSymbol}>
        <img
          className={style.weatherIcon}
          src={symbolImgPath}
          alt="Weather symbol"
        />
      </div>
      <Link to={destination.toLowerCase()} className={style.details}>
        <hr className={style.destinationCardDivider} />
        <p>Detaljer</p>
      </Link>
    </div>
  );
}
