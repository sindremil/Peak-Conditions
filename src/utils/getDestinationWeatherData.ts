import { useWeatherDataQuery } from '../hooks/useWeatherData';
import DestinationPoint from '../interfaces/DestinationPoint';
import SelectedWeatherData from '../interfaces/SelectedWeatherData';
import WeatherData from '../interfaces/WeatherData';
import isValidWeatherData from './isValidWeatherData';

export default function getDestinationWeatherData(
  destinationName: string,
  point: number
): WeatherData {
  const destinationPoint: DestinationPoint = {
    destination: destinationName,
    pointIndex: point,
  };
  const { data, isLoading, isError, isFetched, isRefetching, isSuccess } =
    useWeatherDataQuery(destinationPoint);

  const weatherData: WeatherData = {
    type: '',
    geometry: {
      type: '',
      coordinates: [0, 0, 0],
    },
    properties: {
      meta: {
        updated_at: '',
        units: {
          air_pressure_at_sea_level: '',
          air_temperature: '',
          cloud_area_fraction: '',
          precipitation_amount: '',
          relative_humidity: '',
          wind_from_direction: '',
          wind_speed: '',
        },
      },
      timeseries: [],
    },
  };

  if (isLoading) {
    console.log('Loading weather data');
    return weatherData;
  } else if (isError) {
    console.log('Error loading weather data');
    return weatherData;
  } else if (isFetched && isSuccess && !isRefetching) {
    return data;
  } else {
    return weatherData;
  }
}

export function getSelectedWeatherData(
  destinationName: string,
  point: number,
  timeseries: number
): SelectedWeatherData {
  const data = getDestinationWeatherData(destinationName, point);

  let weatherData: SelectedWeatherData = {
    destination: '',
    temperature: 0,
    windSpeed: 0,
    symbolCode: '',
  };

  if (isValidWeatherData(data)) {
    const dataPath = data.properties.timeseries[timeseries].data;
    weatherData = {
      destination: destinationName,
      temperature: dataPath.instant.details.air_temperature,
      windSpeed: dataPath.instant.details.wind_speed,
      symbolCode: dataPath.next_1_hours.summary.symbol_code,
    };
  }
  return weatherData;
}
