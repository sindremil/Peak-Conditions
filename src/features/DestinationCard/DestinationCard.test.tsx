import { describe, expect, it } from 'vitest';
<<<<<<< HEAD
import DestinationCard from '../DestinationCard/DestinationCard';
=======
import DestinationCard from './DestinationCard';
>>>>>>> a84b5a8e468e638542677cc51adf5a205c88b6dc
import { render, screen, fireEvent } from '@testing-library/react';
import SelectedWeatherData from '../../interfaces/SelectedWeatherData';
import { MemoryRouter } from 'react-router-dom';
import notFavourite from './../assets/star1.svg';
import favourite from './../assets/star2.svg';
import { addFavourite, removeFavourite, isFavourite } from '../../utils/favourite';
import renderer from 'react-test-renderer';

interface MockDestinationCardProps extends SelectedWeatherData {
  isLocalStorageFavourite: boolean;
}

const mockData: MockDestinationCardProps = {
  destination: 'Åre',
  temperature: -3.7,
  windSpeed: 12.1,
  symbolCode: 'heavyrainandthunder',
  isLocalStorageFavourite: false,
};

describe('DestinationCard', () => {
  it('Destination image is visible', () => {
    renderCard();
    const img = screen.getByRole('img', { name: 'Åre' });
    expect(img).toBeVisible();
  });

  it('Favourite star changes when clicked', () => {
    renderCard();

    /* Check correct initial state (not favourite) */
    const star = screen.getByRole('img', { name: 'Favourite star' });
    expect(!isFavourite('Åre'));
    expect(star).toHaveAttribute('src', notFavourite);

    /* Click star and check for state change */
    fireEvent.click(star);
    addFavourite('Åre');
    expect(isFavourite('Åre'));
    expect(star).toHaveAttribute('src', favourite);

    /* Click star again to check state change back */
    fireEvent.click(star);
    removeFavourite('Åre');
    expect(!isFavourite('Åre'));
    expect(star).toHaveAttribute('src', notFavourite);
  });

  it('DestinationCard snapshot', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <DestinationCard
            destination={mockData.destination}
            temperature={mockData.temperature}
            windSpeed={mockData.windSpeed}
            symbolCode={mockData.symbolCode}
            isLocalStorageFavourite={mockData.isLocalStorageFavourite}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

/* Renders a destination card with mock data */
function renderCard() {
  render(
    <MemoryRouter>
      <DestinationCard
        destination={mockData.destination}
        temperature={mockData.temperature}
        windSpeed={mockData.windSpeed}
        symbolCode={mockData.symbolCode}
        isLocalStorageFavourite={mockData.isLocalStorageFavourite}
      />
    </MemoryRouter>
  );
}
