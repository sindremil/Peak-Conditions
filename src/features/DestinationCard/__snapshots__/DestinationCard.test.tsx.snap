// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DestinationCard > DestinationCard snapshot 1`] = `
<div
  className="card destinationCard"
>
  <div
    className="imgContainer"
  >
    <img
      alt="Åre"
      className="destinationCardImg"
      src="src/assets/destinationimages/Åre.jpg"
    />
    <img
      alt="Favourite star"
      className="favourite"
      onClick={[Function]}
      src="/project1/src/assets/star1.svg"
    />
  </div>
  <div
    className="destinationNameContainer"
  >
    <p
      className="destinationName"
    >
      Åre
    </p>
  </div>
  <div
    className="destinationInfoContainer"
  >
    <p
      className="destinationTemp"
    >
      -3.7
      °
    </p>
    <p
      className="destinationWind"
    >
      12.1
      m/s
    </p>
  </div>
  <div
    className="destinationSymbol"
  >
    <img
      className="weatherIcon"
      src="https://raw.githubusercontent.com/metno/weathericons/89e3173756248b4696b9b10677b66c4ef435db53/weather/svg/heavyrainandthunder.svg"
    />
  </div>
  <a
    className="details"
    href="/åre"
    onClick={[Function]}
  >
    <hr
      className="destinationCardDivider"
    />
    <p>
      Detaljer
    </p>
  </a>
</div>
`;
